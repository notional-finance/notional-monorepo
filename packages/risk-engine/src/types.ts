import { TokenBalance, TokenDefinition } from '@notional-finance/core-entities';

/** Denomination can be either a specific currency id or symbol */
export type SymbolOrID = string | number;

export type RiskFactorKeys = keyof RiskFactors;

export interface RiskFactors {
  // Absolute values
  netWorth: () => TokenBalance;
  freeCollateral: () => TokenBalance;
  // Ratios
  loanToValue: () => number;
  collateralRatio: () => number | null;
  healthFactor: () => number | null;
  leverageRatio: (currencyId?: number) => number | null;
  // Thresholds
  liquidationPrice: (
    debt: TokenDefinition,
    collateral: TokenDefinition
  ) => TokenBalance | null;
  assetLiquidationThreshold: (asset: TokenDefinition) => TokenBalance | null;
}

export interface RiskFactorLimit<F extends RiskFactorKeys> {
  riskFactor: F;
  limit: NonNullable<ReturnType<RiskFactors[F]>>;
  args?: Parameters<RiskFactors[F]>;
}
