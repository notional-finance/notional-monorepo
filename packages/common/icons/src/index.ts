export * from './lib/analytics-icon/analytics-icon';
export * from './lib/arbitrum-icon/arbitrum-icon';
export * from './lib/stack-icon/stack-icon';
export * from './lib/stake-icon/stake-icon';
export * from './lib/token-icon/token-icon';
export * from './lib/token-terminal-icon/token-terminal-icon';
export * from './lib/vault-icon/vault-icon';
export * from './lib/governance-icon/governance-icon';
export * from './lib/history-icon/history-icon';
export * from './lib/ical-icon/ical-icon';
export * from './lib/info-icon/info-icon';
export * from './lib/docs-icon/docs-icon';
export * from './lib/download-icon/download-icon';
export * from './lib/pie-chart-icon/pie-chart-icon';
export * from './lib/coins-icon/coins-icon';
export * from './lib/coins-circle-icon/coins-circle-icon';
export * from './lib/copy-icon/copy-icon';
export * from './lib/bar-chart-icon/bar-chart-icon';
export * from './lib/bar-chart-lateral-icon/bar-chart-lateral-icon';
export * from './lib/bar-chart-lightning-icon/bar-chart-lightning-icon';
export * from './lib/portfolio-chart-icon/portfolio-chart-icon';
export * from './lib/launch-icon/launch-icon';
export * from './lib/lightning-icon/lightning-icon';
export * from './lib/microsoft-cal-icon/microsoft-cal-icon';
export * from './lib/money-market-icon/money-market-icon';
export * from './lib/arrow-icon/arrow-icon';
export * from './lib/arrow-right-alt-icon/arrow-right-alt-icon';
export * from './lib/arrow-right-icon/arrow-right-icon';
export * from './lib/usd-icon/usd-icon';
export * from './lib/jpy-icon/jpy-icon';
export * from './lib/cad-icon/cad-icon';
export * from './lib/gbp-icon/gbp-icon';
export * from './lib/chf-icon/chf-icon';
export * from './lib/eur-icon/eur-icon';
export * from './lib/resources-icon/resources-icon';
export * from './lib/signature-icon/signature-icon';
export * from './lib/sparkles-icon/sparkles-icon';
export * from './lib/notional-icon/notional-icon';
export * from './lib/notional-plain-icon/notional-plain-icon';
export * from './lib/outlook-cal-icon/outlook-cal-icon';
export * from './lib/discord-icon/discord-icon';
export * from './lib/discord-footer/discord-footer';
export * from './lib/discord-plain-icon/discord-plain-icon';
export * from './lib/blocks-icon/blocks-icon';
export * from './lib/dashboard-icon/dashboard-icon';
export * from './lib/insurance-icon/insurance-icon';
export * from './lib/international-icon/international-icon';
export * from './lib/japanese-flag-icon/japanese-flag-icon';
export * from './lib/bug-bounty-icon/bug-bounty-icon';
export * from './lib/cal-add-icon/cal-add-icon';
export * from './lib/developer-docs-icon/developer-docs-icon';
export * from './lib/defi-Llama-icon/defi-Llama-icon';
export * from './lib/degen-score-icon/degen-score-icon';
export * from './lib/dune-icon/dune-icon';
export * from './lib/forum-icon/forum-icon';
export * from './lib/equals-icon/equals-icon';
export * from './lib/four-square-icon/four-square-icon';
export * from './lib/gear-icon/gear-icon';
export * from './lib/google-cal-icon/google-cal-icon';
export * from './lib/checkmark-icon/checkmark-icon';
export * from './lib/checkmark-round/checkmark-round-icon';
export * from './lib/chevron-down-icon/chevron-down-icon';
export * from './lib/chinese-flag-icon/chinese-flag-icon';
export * from './lib/circle-icon/circle-icon';
export * from './lib/close-x/close-x';
export * from './lib/swap-icon/swap-icon';
export * from './lib/swap-vertical-icon/swap-vertical-icon';
export * from './lib/angled-arrow-icon/angled-arrow-icon';
export * from './lib/reset-icon/reset-icon';
export * from './lib/external-link-icon/external-link-icon';
export * from './lib/eye-icon/eye-icon';
export * from './lib/filter-icon/filter-icon';
export * from './lib/wave-button-icon/wave-button-icon';
export * from './lib/active-bell-icon/active-bell-icon';
export * from './lib/alert-icon/alert-icon';
export * from './lib/bell-icon/bell-icon';
export * from './lib/v3-beta-icon/v3-beta-icon';
export * from './lib/announcement-icon/announcement-icon';
export * from './lib/plus-icon/plus-icon';
export * from './lib/note-with-shadow/note-with-shadow';
export * from './lib/up-and-down-icon/up-and-down-icon';
export * from './assets/icons/currencies/eth-dai-usdc-wbtc.svg';
export * from './lib/leverage-header-icons/leverage-header-icons';
export * from './lib/krw-icon/krw-icon';
export * from './lib/cny-icon/cny-icon';
export * from './lib/brl-icon/brl-icon';
export * from './lib/note-icon/note-icon';
export * from './lib/eth-icon/eth-icon';
export * from './lib/sgd-icon/sgd-icon';
export * from './lib/try-icon/try-icon';
export * from './lib/double-token-icon/double-token-icon';
export * from './lib/wallet-icon/wallet-icon';
export * from './lib/leaf-icon/leaf-icon';
