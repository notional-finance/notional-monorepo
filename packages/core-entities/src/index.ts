export * from './token-balance';
export * from './Definitions';
export * from './Matchers';
export * from './Registry';
export { AccountFetchMode } from './client/account-registry-client';
export * from './exchanges';
export * from './vaults';
export { ConfigurationClient, NOTERegistryClient } from './client';
export * from './config/fiat-config';
export * from './server';
export { ClientRegistry } from './client/client-registry';
export type { ServerRegistry } from './server/server-registry';
export type { GraphDocument } from './server/analytics-server';
export {
  fetchGraph,
  loadGraphClientDeferred,
  fetchGraphPaginate,
} from './server/server-registry';
export * from './config/whitelisted-vaults';
export { DeprecatedVaults } from './server/vault-overrides';
export { SecondaryIncentiveToken } from './config/whitelisted-tokens';
export * from './Boosts';

export { NetworkClientModel, NetworkServerModel } from './models/NetworkModel';
