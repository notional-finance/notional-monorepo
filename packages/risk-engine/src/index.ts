export * from './types';
export { AccountRiskProfile } from './account-risk';
import { BaseRiskProfile } from './base-risk';

export type RiskProfile = typeof BaseRiskProfile;
