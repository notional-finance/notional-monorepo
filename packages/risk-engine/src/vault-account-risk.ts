import {
  Registry,
  TokenBalance,
  TokenDefinition,
} from '@notional-finance/core-entities';
import {
  Network,
  PRIME_CASH_VAULT_MATURITY,
  RATE_DECIMALS,
  RATE_PRECISION,
  unique,
} from '@notional-finance/util';
import { BaseRiskProfile } from './base-risk';
import { SymbolOrID } from './types';

export class VaultAccountRiskProfile extends BaseRiskProfile {
  static collateralToLeverageRatio(collateralRatio: number) {
    return 1 / collateralRatio;
  }

  static leverageToCollateralRatio(leverageRatio: number) {
    return 1 / leverageRatio;
  }

  static empty(network: Network, vaultAddress: string, maturity: number) {
    const config = Registry.getConfigurationRegistry();
    const { vaultShareID } = config.getVaultIDs(
      network,
      vaultAddress,
      maturity
    );
    return new VaultAccountRiskProfile(vaultAddress, [
      TokenBalance.fromID(0, vaultShareID, network),
    ]);
  }

  static from(vaultAddress: string, balances: TokenBalance[]) {
    return new VaultAccountRiskProfile(vaultAddress, balances);
  }

  static simulate(
    vaultAddress: string,
    from: TokenBalance[],
    apply: TokenBalance[]
  ) {
    return new VaultAccountRiskProfile(vaultAddress, [...from, ...apply]);
  }

  simulate(apply: TokenBalance[]) {
    return VaultAccountRiskProfile.simulate(
      this.vaultAddress,
      this.balances,
      apply
    );
  }

  public discountFCash: boolean;

  /** Takes a set of token balances to create a new vault account risk profile */
  constructor(public vaultAddress: string, _balances: TokenBalance[]) {
    const balances = _balances.filter(
      (t) => t.isVaultToken && t.token.vaultAddress === vaultAddress
    );
    const network = unique(balances.map((b) => b.token.network));
    const maturity = unique(balances.map((b) => b.token.maturity));

    // If initiating a new vault position, at least the dummy value of zero vault
    // shares in the proper maturity must be passed in.
    if (network.length != 1 || maturity.length != 1)
      throw Error('All balances must be in same vault, network and maturity');

    const denom = balances.find((b) => b.tokenType === 'VaultShare')?.token
      .underlying;
    if (!denom) throw Error('Underlying not defined');

    // NOTE: this will settle balances inside
    super(balances, denom);

    this.discountFCash =
      Registry.getConfigurationRegistry().getVaultDiscountfCash(
        network[0],
        this.vaultAddress
      );
  }

  get vaultConfig() {
    return Registry.getConfigurationRegistry().getVaultConfig(
      this.network,
      this.vaultAddress
    );
  }

  get vaultLeverageFactors() {
    return Registry.getConfigurationRegistry().getVaultLeverageFactors(
      this.network,
      this.vaultAddress
    );
  }

  get maturity() {
    return this.vaultShares.maturity;
  }

  get vaultDebt() {
    return (
      this.debts.find((t) => t.tokenType === 'VaultDebt') ||
      TokenBalance.zero(
        Registry.getTokenRegistry().getVaultDebt(
          this.network,
          this.vaultAddress,
          this.maturity
        )
      )
    );
  }

  get vaultCash() {
    return (
      this.balances.find((t) => t.tokenType === 'VaultCash') ||
      TokenBalance.zero(
        Registry.getTokenRegistry().getVaultCash(
          this.network,
          this.vaultAddress,
          this.maturity
        )
      )
    );
  }

  get vaultAdapter() {
    return Registry.getVaultRegistry().getVaultAdapter(
      this.network,
      this.vaultAddress
    );
  }

  get maxLeverageRatio() {
    return VaultAccountRiskProfile.collateralToLeverageRatio(
      this.vaultConfig.minCollateralRatioBasisPoints / RATE_PRECISION
    );
  }

  get vaultShares() {
    const v = this.balances.find((t) => t.tokenType === 'VaultShare');
    if (!v) throw Error('Vault Shares not found');
    return v;
  }

  protected _netCurrencyDebt() {
    return this.allCurrencyIds.map((id) => {
      const underlying = Registry.getTokenRegistry().getUnderlying(
        this.network,
        id
      );
      const zeroUnderlying = TokenBalance.zero(underlying);

      const debt =
        this.balances
          .find((t) => t.token.currencyId === id && t.tokenType == 'VaultDebt')
          ?.toUnderlying() || zeroUnderlying;

      const cash =
        this.balances
          .find((t) => t.token.currencyId === id && t.tokenType == 'VaultCash')
          ?.toUnderlying() || zeroUnderlying;

      return debt.add(cash);
    });
  }

  totalAssetsRiskAdjusted(denominated = this.defaultSymbol) {
    const denom = this.denom(denominated);
    const vaultShareValue = this.vaultShares.toUnderlying();

    return this._netCurrencyDebt()
      .map((netDebt) => {
        return netDebt.isPositive()
          ? netDebt.toToken(denom)
          : TokenBalance.from(0, denom);
      })
      .reduce(VaultAccountRiskProfile._sum, vaultShareValue);
  }

  /** Total debt with risk adjustments */
  totalDebtRiskAdjusted(denominated = this.defaultSymbol) {
    const denom = this.denom(denominated);

    return this._netCurrencyDebt()
      .map((netDebt) => {
        return netDebt.isPositive()
          ? TokenBalance.zero(denom)
          : netDebt.toToken(denom);
      })
      .reduce(VaultAccountRiskProfile._sum, TokenBalance.zero(denom));
  }

  collateralRatio(): number | null {
    const totalDebt = this.totalDebtRiskAdjusted().neg();
    const totalAssets = this.totalAssetsRiskAdjusted();
    return totalDebt.isZero()
      ? null
      : totalAssets.sub(totalDebt).ratioWith(totalDebt).toNumber() /
          RATE_PRECISION;
  }

  assetLiquidationThreshold(
    // NOTE: this parameter is unused because the returned units is always in vault share denomination
    _collateral: TokenDefinition = this.vaultShares.token
  ): TokenBalance | null {
    if (this.vaultShares.isZero() || this.vaultDebt.isZero()) return null;

    // (minCollateralRatio + 1) * debtOutstanding = vaultSharesValue
    const config = Registry.getConfigurationRegistry();
    const vaultConfig = config.getVaultConfig(this.network, this.vaultAddress);
    // NOTE: this value is in primary borrow underlying terms
    const oneVaultShareValueAtLiquidation = this.totalDebtRiskAdjusted()
      .neg()
      .scale(
        vaultConfig.minCollateralRatioBasisPoints + RATE_PRECISION,
        this.vaultShares.scaleTo(RATE_DECIMALS)
      );

    // This is the relative exchange rate decrease of vault shares to liquidation
    return oneVaultShareValueAtLiquidation.toToken(this.vaultShares.token);
  }

  freeCollateral(): TokenBalance {
    throw new Error('Method not implemented.');
  }

  healthFactor(): number | null {
    throw new Error('Method not implemented.');
  }

  netCollateralAvailable(_collateral: SymbolOrID): TokenBalance {
    return this.totalAssetsRiskAdjusted().sub(this.totalDebtRiskAdjusted());
  }

  leverageRatio() {
    const collateralRatio = this.collateralRatio();
    if (collateralRatio) {
      return 1 / collateralRatio;
    } else if (collateralRatio === 0) {
      return Infinity;
    } else {
      return null;
    }
  }

  aboveMaxLeverageRatio() {
    const leverage = this.leverageRatio();
    return (
      leverage !== null && leverage > this.vaultLeverageFactors.maxLeverageRatio
    );
  }

  getAllRiskFactors() {
    return {
      netWorth: this.netWorth(),
      debts: this.totalDebt(),
      assets: this.totalAssets(),
      collateralRatio: this.collateralRatio(),
      liquidationPrice: this.getAllLiquidationPrices({
        onlyUnderlyingDebt: true,
      }),
      aboveMaxLeverageRatio: this.aboveMaxLeverageRatio(),
      leverageRatio: this.leverageRatio(),
      // TODO: this also needs to include debt....
      assetLiquidationThreshold: this.collateral.map((a) =>
        this.assetLiquidationThreshold(a.token)
      ),
    };
  }

  override maxWithdraw(
    _token: TokenDefinition = this.vaultShares.token
  ): TokenBalance {
    let costToRepay: TokenBalance | undefined;
    if (this.vaultDebt.maturity !== PRIME_CASH_VAULT_MATURITY) {
      const fCash = Registry.getExchangeRegistry().getfCashMarket(
        this.network,
        this.vaultDebt.currencyId
      );

      try {
        const { tokensOut } = fCash.calculateTokenTrade(
          this.vaultDebt.unwrapVaultToken(),
          0
        );

        costToRepay = tokensOut
          .add(this.vaultCash.unwrapVaultToken())
          .toUnderlying()
          .neg();
      } catch {
        costToRepay = undefined;
      }
    }

    if (!costToRepay) {
      // If the trading fails or the debt is in prime debt, repay at a
      // 1-1 prime cash price
      costToRepay = this.vaultDebt
        .unwrapVaultToken()
        .toPrimeCash()
        .add(this.vaultCash.unwrapVaultToken())
        .toUnderlying()
        .neg();
    }

    // Vault shares burned to repay debt
    const { netVaultSharesForUnderlying } =
      this.vaultAdapter.getNetVaultSharesMinted(
        costToRepay, // this is a negative number
        this.vaultShares.token
      );

    // Return this in vault shares terms
    return this.vaultShares.gt(netVaultSharesForUnderlying)
      ? this.vaultShares.sub(netVaultSharesForUnderlying)
      : this.vaultShares.copy(0);
  }
}
