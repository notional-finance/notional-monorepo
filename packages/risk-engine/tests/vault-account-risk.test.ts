import {
  AccountFetchMode,
  Registry,
  TokenBalance,
} from '@notional-finance/core-entities';
import { Network, PRIME_CASH_VAULT_MATURITY } from '@notional-finance/util';
import { VaultAccountRiskProfile } from '../src/vault-account-risk';

describe.withForkAndRegistry(
  {
    network: Network.ArbitrumOne,
    fetchMode: AccountFetchMode.SINGLE_ACCOUNT_DIRECT,
  },
  'Vault Risk',
  () => {
    const vaultAddress = '0xae38f4b960f44d86e798f36a374a1ac3f2d859fa';
    const maturity = 1695168000;

    describe('Settled Vault Account', () => {
      const settledMaturity = 1687392000;
      it('converts a settled vault account to prime', () => {
        const config = Registry.getConfigurationRegistry();
        Registry.getTokenRegistry().registerToken(
          config.vaultIdToTokenDefinition(
            config.getVaultIDs(
              Network.ArbitrumOne,
              vaultAddress,
              settledMaturity
            ).primaryCashID,
            Network.ArbitrumOne
          )
        );

        const profile = VaultAccountRiskProfile.from(vaultAddress, [
          TokenBalance.fromFloat(
            10,
            Registry.getTokenRegistry().getVaultShare(
              Network.ArbitrumOne,
              vaultAddress,
              settledMaturity
            )
          ),
          TokenBalance.fromFloat(
            -8,
            Registry.getTokenRegistry().getVaultDebt(
              Network.ArbitrumOne,
              vaultAddress,
              settledMaturity
            )
          ),
          TokenBalance.fromFloat(
            2,
            Registry.getTokenRegistry().getVaultCash(
              Network.ArbitrumOne,
              vaultAddress,
              settledMaturity
            )
          ),
        ]);

        // Vault shares and vault cash convert 1-1
        expect(profile.settledBalances.length).toBe(3);
        expect(profile.vaultShares.maturity).toBe(PRIME_CASH_VAULT_MATURITY);
        expect(profile.vaultShares.toFloat()).toBe(10);
        expect(profile.vaultCash.maturity).toBe(PRIME_CASH_VAULT_MATURITY);
        expect(profile.vaultCash.toFloat()).toBe(2);

        // This converts at a non 1-1 rate.
        expect(profile.vaultDebt.maturity).toBe(PRIME_CASH_VAULT_MATURITY);
        expect(profile.vaultDebt.toUnderlying().toFloat()).toBeLessThan(-8);
      });
    });

    describe('Risk Factors', () => {
      it('calculates empty leverage ratio and collateral ratio', () => {
        const vaultRiskProfile = VaultAccountRiskProfile.empty(
          Network.ArbitrumOne,
          vaultAddress,
          maturity
        );

        expect(vaultRiskProfile.leverageRatio()).toBe(null);
        expect(vaultRiskProfile.collateralRatio()).toBe(null);
      });

      it('calculates no debt leverage ratio and collateral ratio', () => {
        const vaultRiskProfile = VaultAccountRiskProfile.from(vaultAddress, [
          TokenBalance.fromFloat(
            1,
            Registry.getTokenRegistry().getVaultShare(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
        ]);

        expect(vaultRiskProfile.leverageRatio()).toBe(null);
        expect(vaultRiskProfile.collateralRatio()).toBe(null);
      });

      it('calculates debt leverage ratio and collateral ratio', () => {
        const vaultRiskProfile = VaultAccountRiskProfile.from(vaultAddress, [
          TokenBalance.fromFloat(
            1,
            Registry.getTokenRegistry().getVaultShare(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
          TokenBalance.fromFloat(
            -0.9,
            Registry.getTokenRegistry().getVaultDebt(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
        ]);

        const risks = vaultRiskProfile.getAllRiskFactors();
        expect(risks).toMatchSnapshot();
        // expect(collateralRatio).toBeCloseTo(0.125);
        // expect(leverageRatio).toBeCloseTo(7.7109);
        // expect(liquidationPrice[0].price?.toFloat()).toBeCloseTo(0.9725);
        // expect(assetLiquidationThreshold[0]?.toFloat()).toBeCloseTo(0.9648);
      });

      it('calculates debt with cash leverage ratio and collateral ratio', () => {
        const vaultRiskProfile = VaultAccountRiskProfile.from(vaultAddress, [
          TokenBalance.fromFloat(
            1,
            Registry.getTokenRegistry().getVaultShare(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
          TokenBalance.fromFloat(
            -0.9,
            Registry.getTokenRegistry().getVaultDebt(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
          TokenBalance.fromFloat(
            0.01,
            Registry.getTokenRegistry().getVaultCash(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
        ]);

        const risks = vaultRiskProfile.getAllRiskFactors();
        expect(risks).toMatchSnapshot();
      });
    });

    describe('Max Withdraw', () => {
      it('fCash', () => {
        const vaultRiskProfile = VaultAccountRiskProfile.from(vaultAddress, [
          TokenBalance.fromFloat(
            10,
            Registry.getTokenRegistry().getVaultShare(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
          TokenBalance.fromFloat(
            -8,
            Registry.getTokenRegistry().getVaultDebt(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
        ]);

        expect(vaultRiskProfile.maxWithdraw().token.id).toBe(
          vaultRiskProfile.vaultShares.tokenId
        );
        expect(vaultRiskProfile.maxWithdraw().toFloat()).toBeCloseTo(2.094);
      });

      it('Prime Cash', () => {
        const vaultRiskProfile = VaultAccountRiskProfile.from(vaultAddress, [
          TokenBalance.fromFloat(
            10,
            Registry.getTokenRegistry().getVaultShare(
              Network.ArbitrumOne,
              vaultAddress,
              PRIME_CASH_VAULT_MATURITY
            )
          ),
          TokenBalance.fromFloat(
            -8,
            Registry.getTokenRegistry().getVaultDebt(
              Network.ArbitrumOne,
              vaultAddress,
              PRIME_CASH_VAULT_MATURITY
            )
          ),
        ]);

        expect(vaultRiskProfile.maxWithdraw().token.id).toBe(
          vaultRiskProfile.vaultShares.tokenId
        );
        expect(vaultRiskProfile.maxWithdraw().toFloat()).toBeCloseTo(1.959);
      });

      it('fCash + Cash', () => {
        const vaultRiskProfile = VaultAccountRiskProfile.from(vaultAddress, [
          TokenBalance.fromFloat(
            10,
            Registry.getTokenRegistry().getVaultShare(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
          TokenBalance.fromFloat(
            -8,
            Registry.getTokenRegistry().getVaultDebt(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
          TokenBalance.fromFloat(
            2,
            Registry.getTokenRegistry().getVaultCash(
              Network.ArbitrumOne,
              vaultAddress,
              maturity
            )
          ),
        ]);

        expect(vaultRiskProfile.maxWithdraw().token.id).toBe(
          vaultRiskProfile.vaultShares.tokenId
        );
        expect(vaultRiskProfile.maxWithdraw().toFloat()).toBeCloseTo(4.093);
      });
    });
  }
);
