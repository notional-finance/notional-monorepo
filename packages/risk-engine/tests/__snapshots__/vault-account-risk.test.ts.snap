// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Registry Vault Risk Risk Factors calculates debt leverage ratio and collateral ratio 1`] = `
Object {
  "aboveMaxLeverageRatio": false,
  "assetLiquidationThreshold": Array [
    Object {
      "_isTokenBalance": true,
      "hex": "0x05c7103c",
      "network": "arbitrum",
      "tokenId": "0x00000000ae38f4b960f44d86e798f36a374a1ac3f2d859fa000300650a360009",
    },
  ],
  "assets": Object {
    "_isTokenBalance": true,
    "hex": "0x0f61da",
    "network": "arbitrum",
    "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
  },
  "collateralRatio": 0.124500545,
  "debts": Object {
    "_isTokenBalance": true,
    "hex": "-0x0dadde",
    "network": "arbitrum",
    "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
  },
  "leverageRatio": 8.032093353486927,
  "liquidationPrice": Array [
    Object {
      "collateral": Object {
        "address": "0x1344a36a1b56144c3bc62e7757377d288fde0369",
        "currencyId": 3,
        "decimals": 8,
        "id": "0x00000000ae38f4b960f44d86e798f36a374a1ac3f2d859fa000300650a360009",
        "isFCashDebt": false,
        "maturity": 1695168000,
        "name": "Vault Shares in USDC for 0xae38f4b960f44d86e798f36a374a1ac3f2d859fa Fixed Term @ 1695168000",
        "network": "arbitrum",
        "symbol": "vsUSDC:0xae38f4b960f44d86e798f36a374a1ac3f2d859fa:fixed@1695168000",
        "tokenInterface": "ERC1155",
        "tokenType": "VaultShare",
        "totalSupply": Object {
          "_isTokenBalance": true,
          "hex": "0x2ee8204f",
          "network": "arbitrum",
          "tokenId": "0x00000000ae38f4b960f44d86e798f36a374a1ac3f2d859fa000300650a360009",
        },
        "underlying": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "vaultAddress": "0xae38f4b960f44d86e798f36a374a1ac3f2d859fa",
      },
      "debt": Object {
        "address": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "currencyId": 3,
        "decimals": 6,
        "id": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "isFCashDebt": false,
        "name": "USD Coin (Arb1)",
        "network": "arbitrum",
        "symbol": "USDC",
        "tokenInterface": "ERC20",
        "tokenType": "Underlying",
      },
      "price": Object {
        "_isTokenBalance": true,
        "hex": "0x0ee908",
        "network": "arbitrum",
        "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
      },
      "riskExposure": Object {
        "isCrossCurrency": false,
        "isPrimeDebt": false,
        "risk": "VaultShare",
      },
    },
  ],
  "netWorth": Object {
    "_isTokenBalance": true,
    "hex": "0x01b3fc",
    "network": "arbitrum",
    "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
  },
}
`;

exports[`With Registry Vault Risk Risk Factors calculates debt with cash leverage ratio and collateral ratio 1`] = `
Object {
  "aboveMaxLeverageRatio": false,
  "assetLiquidationThreshold": Array [
    Object {
      "_isTokenBalance": true,
      "hex": "0x05b66da4",
      "network": "arbitrum",
      "tokenId": "0x00000000ae38f4b960f44d86e798f36a374a1ac3f2d859fa000300650a360009",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x05b66da4",
      "network": "arbitrum",
      "tokenId": "0x00000000ae38f4b960f44d86e798f36a374a1ac3f2d859fa000300650a360009",
    },
  ],
  "assets": Object {
    "_isTokenBalance": true,
    "hex": "0x0f893c",
    "network": "arbitrum",
    "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
  },
  "collateralRatio": 0.137290781,
  "debts": Object {
    "_isTokenBalance": true,
    "hex": "-0x0dadde",
    "network": "arbitrum",
    "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
  },
  "leverageRatio": 7.283810265454022,
  "liquidationPrice": Array [
    Object {
      "collateral": Object {
        "address": "0x1344a36a1b56144c3bc62e7757377d288fde0369",
        "currencyId": 3,
        "decimals": 8,
        "id": "0x00000000ae38f4b960f44d86e798f36a374a1ac3f2d859fa000300650a360009",
        "isFCashDebt": false,
        "maturity": 1695168000,
        "name": "Vault Shares in USDC for 0xae38f4b960f44d86e798f36a374a1ac3f2d859fa Fixed Term @ 1695168000",
        "network": "arbitrum",
        "symbol": "vsUSDC:0xae38f4b960f44d86e798f36a374a1ac3f2d859fa:fixed@1695168000",
        "tokenInterface": "ERC1155",
        "tokenType": "VaultShare",
        "totalSupply": Object {
          "_isTokenBalance": true,
          "hex": "0x2ee8204f",
          "network": "arbitrum",
          "tokenId": "0x00000000ae38f4b960f44d86e798f36a374a1ac3f2d859fa000300650a360009",
        },
        "underlying": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "vaultAddress": "0xae38f4b960f44d86e798f36a374a1ac3f2d859fa",
      },
      "debt": Object {
        "address": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "currencyId": 3,
        "decimals": 6,
        "id": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "isFCashDebt": false,
        "name": "USD Coin (Arb1)",
        "network": "arbitrum",
        "symbol": "USDC",
        "tokenInterface": "ERC20",
        "tokenType": "Underlying",
      },
      "price": Object {
        "_isTokenBalance": true,
        "hex": "0x0ebe1a",
        "network": "arbitrum",
        "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
      },
      "riskExposure": Object {
        "isCrossCurrency": false,
        "isPrimeDebt": false,
        "risk": "VaultShare",
      },
    },
    Object {
      "collateral": Object {
        "address": "0x1344a36a1b56144c3bc62e7757377d288fde0369",
        "currencyId": 3,
        "decimals": 8,
        "id": "0x00000000ae38f4b960f44d86e798f36a374a1ac3f2d859fa000300650a36000b",
        "isFCashDebt": false,
        "maturity": 1695168000,
        "name": "Vault USDC Cash for 0xae38f4b960f44d86e798f36a374a1ac3f2d859fa Fixed Term @ 1695168000",
        "network": "arbitrum",
        "symbol": "vcUSDC:0xae38f4b960f44d86e798f36a374a1ac3f2d859fa:fixed@1695168000",
        "tokenInterface": "ERC1155",
        "tokenType": "VaultCash",
        "underlying": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "vaultAddress": "0xae38f4b960f44d86e798f36a374a1ac3f2d859fa",
      },
      "debt": Object {
        "address": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "currencyId": 3,
        "decimals": 6,
        "id": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
        "isFCashDebt": false,
        "name": "USD Coin (Arb1)",
        "network": "arbitrum",
        "symbol": "USDC",
        "tokenInterface": "ERC20",
        "tokenType": "Underlying",
      },
      "price": Object {
        "_isTokenBalance": true,
        "hex": "0x0ebe1a",
        "network": "arbitrum",
        "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
      },
      "riskExposure": Object {
        "isCrossCurrency": false,
        "isPrimeDebt": false,
        "risk": "VaultCash",
      },
    },
  ],
  "netWorth": Object {
    "_isTokenBalance": true,
    "hex": "0x01db5e",
    "network": "arbitrum",
    "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
  },
}
`;
